pythran/analyses/range_values.py

```diff
@@ -0,0 +1,448 @@
+        """ Update range value for augassigned variables.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = 2; a -= 1")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=1, high=2)
+        """
+        if isinstance(node.target, ast.Name):
+            low, high = self.combine(self.result[node.target.id],
+                                     self.visit(node.value),
+                                     node.op)
+            self.add(node.target.id, low, high)
+
+    def visit_For(self, node):
+        """ Handle iterate variable in for loops.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse('''
+        ... def foo():
+        ...     a = b = c = 2
+        ...     for i in range(1):
+        ...         a -= 1
+        ...         b += 1''')
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=2)
+        >>> res['b']
+        Range(low=2, high=inf)
+        >>> res['c']
+        Range(low=2, high=2)
+        """
+        assert isinstance(node.target, ast.Name), "For apply on variables."
+        self.add(node.target.id, *self.visit(node.iter))
+        old_range = copy.deepcopy(self.result)
+        map(self.visit, node.body)
+        for name, (low, high) in old_range.iteritems():
+            if self.result[name].low < low:
+                self.result[name] = \
+                    self.result[name]._replace(low=-float("inf"))
+            if self.result[name].high > high:
+                self.result[name] = \
+                    self.result[name]._replace(high=float("inf"))
+        map(self.visit, node.orelse)
+
+    def visit_While(self, node):
+        """ Handle incremented variables in loop body.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse('''
+        ... def foo():
+        ...     a = b = c = 2
+        ...     while a > 0:
+        ...         a -= 1
+        ...         b += 1''')
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=2)
+        >>> res['b']
+        Range(low=2, high=inf)
+        >>> res['c']
+        Range(low=2, high=2)
+        """
+        old_range = copy.deepcopy(self.result)
+        map(self.visit, node.body)
+        for name, (low, high) in old_range.iteritems():
+            if self.result[name].low < low:
+                self.result[name] = \
+                    self.result[name]._replace(low=-float("inf"))
+            if self.result[name].high > high:
+                self.result[name] = \
+                    self.result[name]._replace(high=float("inf"))
+        map(self.visit, node.orelse)
```

Comment:
On Tue, Jan 13, 2015 at 01:25:56PM -0800, pbrunet wrote:
> In pythran/analyses/range_values.py:
> > +        Range(low=-inf, high=2)
> > +        >>> res['b']
> > +        Range(low=2, high=inf)
> > +        >>> res['c']
> > +        Range(low=2, high=2)
> > +        &quot;&quot;&quot;
> > +        old_range = copy.deepcopy(self.result)
> > +        map(self.visit, node.body)
> > +        for name, (low, high) in old_range.iteritems():
> > +            if self.result[name].low < low:
> > +                self.result[name] = \
> > +                    self.result[name]._replace(low=-float(&quot;inf&quot;))
> > +            if self.result[name].high > high:
> > +                self.result[name] = \
> > +                    self.result[name]._replace(high=float(&quot;inf&quot;))
> > +        map(self.visit, node.orelse)
> I think the equation is correct but I don't understand how condition will
> modify it?

it's correct, but not as accurate as it could be.